// @peerlancers/json-serialization/dist/spec/json-object.model.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const json_property_1 = require("../src/json-property");
var Gender;
(function (Gender) {
    Gender[Gender["Male"] = 1] = "Male";
    Gender[Gender["Female"] = 2] = "Female";
})(Gender = exports.Gender || (exports.Gender = {}));
class GenderSerialization {
    serialize(value) {
        return `${Gender[value]}`;
    }
    deserialize(value) {
        return Gender[value];
    }
}
exports.GenderSerialization = GenderSerialization;
class Hobby {
    constructor() {
        this.hobbyName = undefined;
        this.description = undefined;
    }
}
__decorate([
    json_property_1.JsonProperty({ name: 'title' }),
    __metadata("design:type", String)
], Hobby.prototype, "hobbyName", void 0);
__decorate([
    json_property_1.JsonProperty(),
    __metadata("design:type", String)
], Hobby.prototype, "description", void 0);
exports.Hobby = Hobby;
class Person {
    constructor() {
        this.firstName = undefined;
        this.lastName = undefined;
        this.currentAge = undefined;
        this.children = undefined;
        this.skills = undefined;
        this.hobbies = undefined;
        this.friends = undefined;
        this.gender = undefined;
        this.ignoredField = undefined;
        this.parent = undefined;
        this.maritalStatus = null;
    }
    get fullName() {
        return `${this.firstName} ${this.lastName}`;
    }
}
__decorate([
    json_property_1.JsonProperty(),
    __metadata("design:type", String)
], Person.prototype, "firstName", void 0);
__decorate([
    json_property_1.JsonProperty(),
    __metadata("design:type", String)
], Person.prototype, "lastName", void 0);
__decorate([
    json_property_1.JsonProperty({ name: 'age' }),
    __metadata("design:type", Number)
], Person.prototype, "currentAge", void 0);
__decorate([
    json_property_1.JsonProperty({ target: Person }),
    __metadata("design:type", Array)
], Person.prototype, "children", void 0);
__decorate([
    json_property_1.JsonProperty(),
    __metadata("design:type", Array)
], Person.prototype, "skills", void 0);
__decorate([
    json_property_1.JsonProperty({ target: Hobby }),
    __metadata("design:type", Array)
], Person.prototype, "hobbies", void 0);
__decorate([
    json_property_1.JsonProperty(),
    __metadata("design:type", Array)
], Person.prototype, "friends", void 0);
__decorate([
    json_property_1.JsonProperty({
        serializer: GenderSerialization,
        deserializer: GenderSerialization
    }),
    __metadata("design:type", Number)
], Person.prototype, "gender", void 0);
__decorate([
    json_property_1.JsonProperty(),
    __metadata("design:type", Person)
], Person.prototype, "parent", void 0);
__decorate([
    json_property_1.JsonProperty(),
    __metadata("design:type", String)
], Person.prototype, "maritalStatus", void 0);
exports.Person = Person;
